### PR TITLE
IH-662: Get rid of non-threadsafe Re.Str usage in XAPI

### PR DESCRIPTION
**Pull request by last-genius**
`Re.Str`, imitating `Str`'s inherently non-threadsafe interface, is also not threadsafe, already in OCaml 4. There are several usage sites left after the [cleanup](https://github.com/xapi-project/xen-api/pull/4639), with most of them in single-threaded CLI tools. One was introduced in [`helpers.ml`](https://github.com/xapi-project/xen-api/pull/4635#pullrequestreview-894893282).

This PR moves to use `Re.Pcre` instead, which has a slightly different interface. Tests were added to ensure the correctness of the new approach (these could be expanded if anybody comes up with another edge case).

### CODE REVIEW

File: ocaml/xapi/helpers.ml [hunk 1]
```diff
@@ -42,21 +42,15 @@ let log_exn_continue msg f x =
...
 let filter_args args =
   List.map
     (fun arg ->
       List.fold_left
         (fun acc (r, t) ->
-          if Re.Str.string_match r acc 0 then
-            Re.Str.replace_matched t acc
-          else
-            acc
+          try String.concat "" [(Re.Pcre.extract ~rex:r acc).(1); t]
```
Comment (edwintorok) [inline ocaml/xapi/helpers.ml:52]: There should be an API that doesn't rely on exceptions, maybe in Re itself

Comment (edwintorok) [reply ocaml/xapi/helpers.ml:52]: There is a `pmatch` function.

File: ocaml/xapi/helpers.ml [hunk 2]
```diff
@@ -42,21 +42,15 @@ let log_exn_continue msg f x =
 type log_output = Always | Never | On_failure
 
 let filter_patterns =
-  [
-    ( Re.Str.regexp "^\\(.*proxy_\\(username\\|password\\)=\\)\\(.*\\)$"
-    , "\\1(filtered)"
-    )
-  ]
+  [(Re.Pcre.regexp "^(.*proxy_(username|password)=)(.*)$", "(filtered)")]
```
Comment (edwintorok) [inline ocaml/xapi/helpers.ml:45]: pcre is fine, though I'm more familiar with the POSIX syntax in Re.Posix.

File: ocaml/xapi/helpers.ml (hunk 1 above)
Comment (last-genius) [reply ocaml/xapi/helpers.ml:52]: Here I am able to do "does it match, and if so, return matching groups" in one go. Switching to `pmatch` means we would take two matches to do the same.

File: ocaml/xapi/helpers.ml (hunk 2 above)
Comment (Vincent-lau) [inline ocaml/xapi/helpers.ml:45]: so with this change you are relying on the fact that `t` consists of `1` and `(filtered)`, right? What if the `t` is something more complicated like `xxx\1yy\2zz\4`, it seems that this would make the t less expressive?

**Comment by Vincent-lau** [comment]
How do we know that Re.Pcre is thread safe?

**Comment by last-genius** [comment]
> How do we know that Re.Pcre is thread safe?

`Re.Str` holds a global state, accessed by functions which take unit, i.e. by definition only operate on this global state, alongside several others, in a non-threadsafe manner. `Re.Pcre` and `Re.Posix` do not have either (global state/functions taking unit)

**Comment by Vincent-lau** [comment]
> > How do we know that Re.Pcre is thread safe?
> 
> `Re.Str` holds a global state, accessed by functions which take unit, i.e. by definition only operate on this global state, alongside several others, in a non-threadsafe manner. `Re.Pcre` and `Re.Posix` do not have either (global state/functions taking unit)

This is knowledge is obtained by by inspecting the source code of `Re`?

File: ocaml/xapi/helpers.ml (hunk 2 above)
Comment (Vincent-lau) [reply ocaml/xapi/helpers.ml:45]: 🆗

**Comment by last-genius** [comment]
> > > How do we know that Re.Pcre is thread safe?
> > 
> > 
> > `Re.Str` holds a global state, accessed by functions which take unit, i.e. by definition only operate on this global state, alongside several others, in a non-threadsafe manner. `Re.Pcre` and `Re.Posix` do not have either (global state/functions taking unit)
> 
> This is knowledge is obtained by by inspecting the source code of `Re`?

Yes, sadly

**Review by Vincent-lau — APPROVED** [review]

File: ocaml/xapi/helpers.ml (hunk 1 above)
Comment (last-genius) [reply ocaml/xapi/helpers.ml:52]: ping @edwintorok

**Comment by psafont** [comment]
This needs some conflicts resolved

**Review by psafont — APPROVED** [review]
